File: Algorithms/24.cpp
/*
#---->@---->@---->@---->@---->@---->@
            ^     ^
            pre   cur
1. pre -> next = cur -> next
             ---------
            /         \
#---->@---->@     @---->@---->@---->@
            ^     ^
            pre   cur
2. pre = pre -> next
             ---------
            /         \
#---->@---->@     @---->@---->@---->@
                  ^     ^
                  cur   pre
3. cur -> next = pre -> next
             ----------
            /          \
#---->@---->@     @     @---->@---->@
                  \          /
                   ----------
                  ^     ^
                  cur   pre
4. pre -> next = cur
             ----------
            /          \
#---->@---->@     @<----@     @---->@
                  \          /
                   ----------
                  ^     ^
                  cur   pre
5. pre = cur; cur = cur -> next
             ---------
            /         \
#---->@---->@     @<----@     @---->@
                  \          /
                   ----------
                  ^           ^
                  pre        cur
*/
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
  ListNode* swapPairs(ListNode* head) {
    if(!head) return NULL;
    ListNode tmp(0);
    tmp.next = head;
    ListNode *pre = &tmp, *cur = head;
    while(cur && cur -> next) {
      pre -> next = cur -> next;
      pre = pre -> next;
      cur -> next = pre -> next;
      pre -> next = cur;
      pre = cur;
      cur = cur -> next;
    }
    return tmp.next;
  }
};

File: Algorithms/24.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def swapPairs(self, head: ListNode) -> ListNode:
        # 1. edge case
        if not head or not head.next:
            return head

        # 2. set two pointers and prev
        dummy = ListNode(val=-1, next=head)
        prev = dummy

        # 3. swap the values
        while head and head.next:
            ptr1, ptr2 = head, head.next

            ptr1.next = ptr2.next
            ptr2.next = ptr1
            prev.next = ptr2

            prev = ptr1
            head = head.next

        return dummy.next
